Updated Get Started with PWAs to use new tooling, new workflow (#669)
MicrosoftDocs/edge-developer: microsoft-edge/progressive-web-apps-chromium/get-started.md

---
description: This guide gives you an overview of PWA basics and tools for building progressive web apps on Windows.
title: Get started with Progressive Web Apps
author: MSEdgeTeam
ms.author: msedgedevrel
ms.date: 05/13/2020
ms.topic: article
ms.prod: microsoft-edge
keywords: progressive web apps, PWA, Edge, Windows, PWABuilder, web manifest, service worker, push
---

# Get started with Progressive Web Apps  

> Progressive Web Apps \(PWAs\) are web apps that are [progressively enhanced][WikiProgressiveEnhancement] with app-like features, such as installation, offline support, and push notifications.  You may also packaged your PWA for app stores, including the Microsoft Store, Google Play, Mac App Store and more.  

Author: I think we can keep Microsoft Store. 2 reasons:

1. The Microsoft Store will still be here for at least another year.
2. We're creating a new path for getting Chromium-based PWAs into the existing Store. We'll have guidance on that likely this month.

That said, I'm flexible. If you feel it's better to highlight AppSource here instead of the Store, I'll change it.

> Progressive Web Apps \(PWAs\) are web apps that are [progressively enhanced][WikiProgressiveEnhancement] with app-like features, such as installation, offline support, and push notifications.  You may also packaged your PWA for app stores, including the Microsoft Store, Google Play, Mac App Store and more.  

Reviewer: I just want to use the correct branding. Do you have a link to the current store?

> Progressive Web Apps \(PWAs\) are web apps that are [progressively enhanced][WikiProgressiveEnhancement] with app-like features, such as installation, offline support, and push notifications.  You may also packaged your PWA for app stores, including the Microsoft Store, Google Play, Mac App Store and more.  

Author: I've updated this to clarify that Microsoft Store = the commercial app store shipped with Windows.

This guide gives you an overview of PWA basics by creating a simple web app and extending it to be a PWA.  The finished product works similarly across modern browsers.  

> [!TIP]
> You may use [PWABuilder][PwaBuilder] to create a new PWA, enhance your existing PWA, or package your PWA for app stores.  

## Prerequisites  

*   Use [VS Code][VisualstudioCodeMain] to edit your PWA source code.  
*   Use [Node.js][NodejsMain] as your local web server.  

## Set up a basic web app  

To get started with an empty web app, generate one using [Node Express App Generator][ExpressjsApplicationGenerator].  Create a new directory, `MySamplePwa` and open a new command-line session \(PowerShell or Bash\).  

In the prompt, run the following commands.  

```shell
npx express-generator --no-view
```  

```shell
npm install
```  

> The commands creates an empty web app and install any dependencies.  

Author: They're required to create the app and get it running. I'd strongly recommend leaving them in.

> The commands creates an empty web app and install any dependencies.  

Author: The linked document contains the full command line documentation, not the specific one we want to use. (in particular, the --no-view option)

> The commands creates an empty web app and install any dependencies.  

Reviewer: Ok that makes sense. Please reject the suggestion.

```shell
npm start
```  

Now browse to `http://localhost:3000` to view your new web app.  

:::image type="complex" source="./media/vs-nodejs-express-index.png" alt-text="Running your new web PWA on localhost":::
   Running your new web PWA on localhost
:::image-end:::

## Turn your app into a PWA  

Now that you have a simple web app, extend it to be a PWA by adding the [3 requirements for PWAs][PwaEdgehtmlIndexRequirements]: [HTTPS](#https), a [Web App Manifest](#web-app-manifest), and a [Service Worker](#service-worker).  

### HTTPS  

Key parts of the PWA platform, such as [Service Workers][MDNServiceWorkerApi], require the use of HTTPS.  When your PWA goes live, you must publish it to an HTTPS URL.  

For debugging purposes, Microsoft Edge also permits `http://localhost` to use the PWA APIs.  

If you [publish this web app as a live site][VisualStudioNodejsTutorialPublishAzureAppService] \(for example, by setting up an [Azure free account][AzureCreateFreeAccount]\), you must ensure your server is configured for HTTPS.  If you use the [Microsoft Azure App Service][AzureWebApps] to host your site, it is served over HTTPS by default.  

> For this guide, use `http://localhost` to build your PWA.  

Author: Fixed.

    var outputArray = new Uint8Array(rawData.length);
    
    for (var i = 0; i < rawData.length; ++i) {
        outputArray[i] = rawData.charCodeAt(i);
    }
    return outputArray;
}
```  

Review the MDN documentation on the [PushManager][MDNPushManager] interface and NPM docs on the [Web-Push][NPMWebPushUsage] library for more details on how the APIs work and various related options.  

### Step 3 - Listen for push notifications  

With your push subscription set up, the remainder of the work happens in the service worker.  First you must set up a handler for server-sent push events, and respond with a toast notification \(if permission was granted\) displaying the push data payload.  Next you add a click handler for the toast to dismiss the notification and sort through a list of currently open windows to open, focus, or open and focus the intended PWA client page.  

In your `pwabuilder-sw.js` file, add the following handlers.  

```javascript
//Respond to a server push with a user notification
self.addEventListener('push', function (event) {
    if (Notification.permission === "granted") {
        const notificationText = event.data.text();

>         const shotNotification = self.registration.showNotification('Sample PWA', {

Author: Good catch. 😊

    console.log('On notification click: ', event.notification.tag);
    event.notification.close();
    
    // This looks to see if the current is already open and focuses it
    event.waitUntil(clients.matchAll({
        type: 'window'
    }).then(function (clientList) {
        for (var i = 0; i < clientList.length; i++) {
            var client = clientList[i];
            if (client.url == 'http://localhost:1337/' && 'focus' in client)
                return client.focus();
        }
        if (clients.openWindow)
            return clients.openWindow('/');
    }));
});
```  

### Step 4 - Try it out  

Time to test push notifications in your PWA!  

1.  Run your PWA in your browser at `localhost:3000`.  
    As your service worker activates and attempts to subscribe your PWA to push notifications, Microsoft Edge asks if you want this PWA to show notifications.  
    
    :::image type="complex" source="./media/notification-permission.png" alt-text="Permission dialog for enabling notifications":::
       Permission dialog for enabling notifications
    :::image-end:::
    
    Select **Allow** to enable notifications.  
    

> 1.  Simulate a server-side push notification.  With your PWA opened at `localhost:3000`, press `F12` to open DevTools.  Go to **Application**, then **Service Worker**, then select **Push** to send a test push notification to your PWA.  You should see a push notification appear near the taskbar.  

Reviewer: ```suggestion
1.  Simulate a server-side push notification.  With your PWA opened at `http://localhost:3000` in your browser, select `F12` to open the DevTools.  Choose **Application** > **Service Worker** > **Push** to send a test push notification to your PWA. You should see a push notification appear near the taskbar.  
```

    
    :::image type="complex" source="./media/devtools-push.png" alt-text="Push a notification from DevTools":::
       Push a notification from DevTools
    :::image-end:::
     

>     If you do not select \(or activate\) a toast notification, it is dismissed after several seconds and queue up in your Windows Action Center.  

Reviewer: ```suggestion
    If you do not select \(or activate\) a toast notification, it is dismissed after several seconds and queues up in your Windows Action Center.  
```

    
    :::image type="complex" source="./media/windows-action-center.png" alt-text="Notifications in Windows Action Center":::
       Notifications in Windows Action Center
    :::image-end:::
    

>     You have the basics of PWA push notifications.  In a real app, the next steps are implemented in a way to manage and store push subscriptions and to properly [encrypt][NPMWebPushEncrypt] payload data being sent across the wire.  

Reviewer: ```suggestion

```

> ## Going further  

Reviewer: ```suggestion
## Next steps
```

> This guide demonstrated the basic anatomy of a Progressive Web App and PWA development tools including Visual Studio Code, PWA Builder, and Microsoft Edge DevTools \(`F12`\).  

Reviewer: ```suggestion

```

> Of course, there is a lot more that goes into [making a great PWA][PwaEdgehtmlIndexRequirements] beyond what you read here, including responsive design, deep-linking, [cross-browser testing][BrowserStackTestEdgeBrowser] and other [best practices][Webhint] \(not to mention your app functionality!\), but hopefully this guide gave you a solid introduction of PWA basics and some ideas on getting started.  If you have further questions on PWA development with Windows or with Visual Studio, please leave a comment!  

Reviewer: ```suggestion
The following is a list of additional tasks to learn when building real-world PWAs:

* Manage and store push subscriptions.
* [Encrypt][NPMWebPushEncrypt] payload data
* Responsive design
* Deep-linking
* [Cross-browser testing][BrowserStackTestEdgeBrowser] 
* Implement best practices such as [Webhint][Webhint]   
```

> Review the other PWA guides to learn how to increase customer engagement and provide a more seamless, OS-integrated app experience.  

Reviewer: ```suggestion

```

> ## See also  
> 
> *   [Progressive Web Apps on MDN web docs][MDNProgressiveWebApps] - Excellent guide on Progressive Web Apps.  
> *   [Progressive Web Apps on web.dev][WebDevProgressiveWebApps] - Excellent guide on Progressive Web Apps.  
> *   [Progressive Web Apps rocks][ProgressiveWebApps] - Showcases real-world examples of PWAs.  
> *   [Hacker News readers as Progressive Web Apps][HackerNewsProgressiveWebApps] - Compares different frameworks and performance patterns for implementing a sample \(Hacker News reader\) PWA.  

Reviewer: ```suggestion
## See also  

*   [Progressive Web Apps on MDN web docs][MDNProgressiveWebApps].  
*   [Progressive Web Apps on web.dev][WebDevProgressiveWebApps].  
*   [Hacker News readers as Progressive Web Apps][HackerNewsProgressiveWebApps] - Compares different frameworks and performance patterns for implementing a sample \(Hacker News reader\) PWA.  
```
